Scrollable Navigation

DenisHdz opened the pull request:
Resolves #2553

**Overall change:**
Add Scrollable Navigation.

**Code changes:**
- Add `CanonicalSwipeableNavigation` and `AmpSwipeableNavigation` components which allow the list to remain on one line under 600px and to be horizontally scrollable.
- Add a gradient overlay as a pseudo-element of the `SwipeableNav`.
- Add `C_POSTBOX_TRANSPARENT` and `C_POSTBOX_OPAQUE` constants to be used in the linear gradient as IE doesn't support 8 digit hex like `{C_POSTBOX}00` and `{C_POSTBOX}FF`.
- Add margin-right/left to the last list item with the same width as the gradient.
- Add custom `useMediaQuery` hook to determine if the document matches the media query `(max-width: 600px)`.
- ~~Add custom `useOverflowed` hook to apply a gradient to the swipeable navigation if the scroll width is bigger than the offset width.~~
- ~~Add custom `useWindowWidth` hook to listen for the window resize and set the width of the window so we can pass it to the `useOverflowed` hook and apply the gradient also on resize.~~
- Add `aria-hidden='true'` to the `CanonicalSwipeableNavigation` under 600px to not expose it to screen readers, as we want that just the links in the Menu dropdown are announced.
- Remove the links from the tab order with `tabindex=-1` under 600px.
- Clone children to pass the `isSwipeable` prop so we can add the `tabindex=-1` under 600px.


---

- [X] I have assigned myself to this PR and the corresponding issues
- [x] Automated jest tests added (for new features) or updated (for existing features)
- [X] This PR requires manual testing

DenisHdz committed:
Add swipeable nav

DenisHdz committed:
Fix conflicts

DenisHdz committed:
Bump version

DenisHdz committed:
Update snapshot

DenisHdz committed:
Remove grid changes from this PR

DenisHdz committed:
Update PR number

DenisHdz committed:
Bump to the right version

DenisHdz committed:
Reemove unkown property

DenisHdz committed:
Update snapshots

DenisHdz committed:
Add direction condition in the gradient

DenisHdz committed:
Remove skip link references from the README

DenisHdz commented on packages/components/psammead-navigation/src/index.jsx:
```diff
@@ -157,11 +179,68 @@ const StyledNav = styled.nav`
+    overflow-x: scroll;
+    white-space: nowrap;
+    scroll-behavior: smooth;
+    -webkit-overflow-scrolling: touch;
```
Tested this on iOS 13 and the navigation doesn't scroll.. If anyone has an idea why, please shout it out.

DenisHdz committed:
Merge branch 'latest' of github.com:bbc/psammead into swipeable-navigation

DenisHdz committed:
Remove overflow hidden from the ul when not mobile

DenisHdz committed:
Update snapshots

DenisHdz committed:
Add and reorganize propTypes, replace isMobile with isSwipeable

DenisHdz committed:
Add border-bottom to each list item above 600px

thekp commented:
![](https://media.giphy.com/media/jKWVkeDxquixAE1gDW/giphy.gif)

Probably worth adding the new the new prop ('isSwipeable') to the NavigationUl readme section.

Also, more of a preference but could the prop type and prop default declarations go at the bottom of the file after the exported components ? then the file structure would be imports>consts>exports>props.

DenisHdz commented:
> Probably worth adding the new the new prop ('isSwipeable') to the NavigationUl readme section.
> 
> Also, more of a preference but could the prop type and prop default declarations go at the bottom of the file after the exported components ? then the file structure would be imports>consts>exports>props.

I agree, but I have moved them after the component declaration as we are trying to do that from now on. Take a look at other components.. happy to discuss it further with all the team.

DenisHdz committed:
Merge branch 'latest' into swipeable-navigation

thekp reviewed (approved):
> I agree, but I have moved them after the component declaration as we are trying to do that from now on. Take a look at other components.. happy to discuss it further with all the team.

if this the case then, LGTM.

thekp committed:
Merge branch 'latest' into swipeable-navigation

DenisHdz committed:
Merge branch 'latest' into swipeable-navigation

DenisHdz committed:
Update README

DenisHdz committed:
Merge branch 'swipeable-navigation' of github.com:bbc/psammead into swipeable-navigation

HarveyPeachey commented:
Just tried on my iPhone 7 and swiping isn't working on the `navigation` story on storybook

HarveyPeachey commented:
> Just tried on my iPhone 7 and swiping isn't working on the `navigation` story on storybook

Actually it does work nvm. It's just the Igbo example has the opaque shadow which makes it look like it's scrollable even though it's at the end of the navigation container:


![ezgif-3-6a4c30d3aadb](https://user-images.githubusercontent.com/4706195/68588616-e0fa1700-0481-11ea-885d-4c5faad37527.gif)

DenisHdz committed:
Merge branch 'latest' into swipeable-navigation

DenisHdz commented:
> > Just tried on my iPhone 7 and swiping isn't working on the `navigation` story on storybook
> 
> Actually it does work nvm. It's just the Igbo example has the opaque shadow which makes it look like it's scrollable even though it's at the end of the navigation container:


Thanks @HarveyPeachey. I am working at the moment on applying the gradient just when the navigation is swipeable 👍

tochwill commented on packages/components/psammead-navigation/src/index.jsx:
```diff
@@ -109,43 +118,91 @@ const CurrentLink = ({ children: link, script, currentPageText }) => (
+        <StyledLink
+          href={url}
+          script={script}
+          currentLink="true"
```
Looking good! I realise this hasn't been changed in this PR, but why is `currentLink` passed as a string instead of a boolean?

DenisHdz replied:
Done

DenisHdz committed:
Add a11y notes

DenisHdz committed:
Merge branch 'swipeable-navigation' of github.com:bbc/psammead into swipeable-navigation

tochwill commented on packages/components/psammead-navigation/src/index.jsx:
```diff
@@ -157,45 +214,73 @@ const StyledNav = styled.nav`
+        ${dir === 'ltr' ? 'right' : 'left'}: 0;
+      `}
+      bottom: 0;
+      z-index: 3;
```
It might be an idea to have z-index constants at some point (something like https://css-tricks.com/handling-z-index/), not for this PR though

tochwill reviewed (commented):
Looks good! If you're working on only adding the gradient when the nav is swipeable I'll take a look again when that's done :)

tochwill commented on packages/components/psammead-navigation/src/index.jsx:
```diff
@@ -157,45 +214,73 @@ const StyledNav = styled.nav`
-  children: node.isRequired,
-};
+const Navigation = ({ children, dir }) => {
+  const [width, setWidth] = useState(window.innerWidth);
```
Hmm, just thinking, how does this work on the server? Is innerWidth undefined at this point?

DenisHdz replied:
I have added a condition so if it is executed on the server-side, the value of the width will be undefined.

DenisHdz committed:
Add hook to check if the navigation is overflowed to apply the gradient

tochwill commented on packages/components/psammead-navigation/src/index.jsx:
```diff
@@ -33,22 +39,25 @@ const StyledUnorderedList = styled.ul`
   position: relative;
-  overflow: hidden;
+
+  ${({ isSwipeable }) => !isSwipeable && `overflow: hidden;`}
```
As `isSwipeable` is set currently set by javascript checking the window's width, could we replace this with a media query here? Then we will get the correct navigation behaviour on the server render as well?

DenisHdz replied:
Good spot 👍

tochwill commented on packages/components/psammead-navigation/src/index.jsx:
```diff
@@ -157,45 +214,103 @@ const StyledNav = styled.nav`
+    const handleResize = () => {
+      setWidth(window.innerWidth);
+    };
+    window.addEventListener('resize', handleResize);
```
Could we replace this with a media query listener? https://developer.mozilla.org/en-US/docs/Web/API/MediaQueryList - Then we can update the value of `isSwipeable` when that event fires, rather than handling every resize event? (Would need to check browser support, I think it's fine!)

DenisHdz replied:
Looking at this now, but I am not sure if we can get rid of the resize event as we need that the `useOverflowed` effect is run every time we resize the window, passing the `width`.

DenisHdz replied:
I have created another custom hook `useMediaQuery` which uses the `matchMedia()` method to determine if the document matches the media query (`max-width: 600px`) https://developer.mozilla.org/en-US/docs/Web/API/Window/matchMedia, so we can then pass the boolean value to `isSwipeable`. 

On the other hand, I am not sure if we can get rid of the resize event as we need that the `useOverflowed` effect is run every time we resize the window, passing the width.

DenisHdz committed:
Pass currentLink as boolean

DenisHdz committed:
Use media query instead

Bopchy reviewed (commented):
The `### When to use this component` section on the README needs updating as the items no longer wrap to the next line

DenisHdz committed:
Update README

DenisHdz committed:
Move hooks to their own folder

DenisHdz committed:
Create SwipeableNavigation folder and tidy main index

DenisHdz committed:
Merge branch 'latest' of github.com:bbc/psammead into swipeable-navigation

DenisHdz committed:
Merge branch 'latest' of github.com:bbc/psammead into swipeable-navigation

DenisHdz committed:
Update snapshots

DenisHdz committed:
Merge branch 'latest' of github.com:bbc/psammead into swipeable-navigation

DenisHdz committed:
Create a new custom hook to a media query listener

DenisHdz committed:
Mock matchMedia

DenisHdz committed:
Merge branch 'latest' into swipeable-navigation

AlistairGempf commented:
![image](https://user-images.githubusercontent.com/43134742/68783423-e185df80-0632-11ea-87ac-67463893a282.png)

The gradient will cover the final item of the navigation item, which makes it look like there is more to swipe to, despite being the end of the list. In the prototype I got around this, I added an extra bit of space at the end of the list, another option is to remove the gradient using javascript when at or near the end of the swipable section.

tochwill commented on packages/components/psammead-navigation/hooks/useWindowWidth.js:
```diff
@@ -0,0 +1,28 @@
+    const handleResize = () => {
+      setWidth(getSize());
+    };
+    window.addEventListener('resize', handleResize);
```
should we debounce this event handling?

DenisHdz replied:
Done. I've added a delay of 100ms.

DenisHdz committed:
Add function to handle the scroll and not apply the gradient when reach the end

DenisHdz committed:
Merge branch 'swipeable-navigation' of github.com:bbc/psammead into swipeable-navigation

DenisHdz committed:
Debounce handleResize with a delay of 100ms

DenisHdz commented:
> ![image](https://user-images.githubusercontent.com/43134742/68783423-e185df80-0632-11ea-87ac-67463893a282.png)
> 
> The gradient will cover the final item of the navigation item, which makes it look like there is more to swipe to, despite being the end of the list. In the prototype I got around this, I added an extra bit of space at the end of the list, another option is to remove the gradient using javascript when at or near the end of the swipable section.

@AlistairGempf I have added a function to handle the scroll and remove the gradient when scrolling to the end of the navigation.

tochwill commented on packages/components/psammead-navigation/hooks/useWindowWidth.js:
```diff
@@ -0,0 +1,30 @@
+
+    const handleResize = () => {
+      setWidth(getSize());
+      console.log('entro');
```
Just remove this :)

DenisHdz replied:
Ups 🙈

DenisHdz committed:
Remove console.log

AlistairGempf commented:
Overall looks really good!

Is there any unit testing we can do of the hooks? I don't know how much is possible and how much benefit it would add, but thought it'd be good to bring it up as a question.

https://stackoverflow.com/questions/45868042/figuring-out-how-to-mock-the-window-size-changing-for-a-react-component-test suggests we can mock changing the window size (although I don't know if it would trigger the 'resize' event.

DenisHdz committed:
Add test for useWindowWidth hook

tochwill commented:
Just had another look! It looks really good, one thing to consider is how is this going to behave on AMP? The majority of functionality is CSS which will work, but the client side JS won't run to check scroll position and screen width

DenisHdz committed:
Add @testing-library/react-hooks and react-test-renderer

DenisHdz committed:
Update useWindowWidth test to use renderHook

DenisHdz committed:
Add test for the useMediaQuery hook

DenisHdz committed:
Add official mocking method for matchMedia

DenisHdz committed:
Add test for the useOverflowed hook

DenisHdz committed:
Merge branch 'latest' into swipeable-navigation

DenisHdz committed:
Compose Swipeable Navigation separately

DenisHdz committed:
Merge branch 'latest' of github.com:bbc/psammead into swipeable-navigation

DenisHdz committed:
Update snapshot

OlgaLyubin commented:
Since you added `react-test-renderer` and `@testing-library/react-hooks` shouldn't the main version of Psammead be bumped up? + CHANGELOG updated

DenisHdz commented:
> Since you added `react-test-renderer` and `@testing-library/react-hooks` shouldn't the main version of Psammead be bumped up? + CHANGELOG updated

Yeah, working on it 👍

OlgaLyubin commented on packages/components/psammead-navigation/src/SwipeableNavigation/index.jsx:
```diff
@@ -0,0 +1,99 @@
+
+const SwipeableNavigation = ({ children, dir }) => {
+  const isSwipeable = useMediaQuery('(max-width: 600px)');
+  const ariaHidden = isSwipeable && { 'aria-hidden': true };
```
As far as I know, for now a non-js navigation is just a swipeable one with dropdown disabled. In that case should we have an option to set `aria-hidden` to false so that tabbing is not disabled? Or is this too much of an exception case that we will not consider?

tochwill replied:
That's something to consider - on the same note, what happens at the moment during SSR? The useMediaQuery hook uses window.matchMedia - does this fall over during server rendering? Storybook just renders on the client doesn't it?

tochwill replied:
We can't know what the width of the page is on server render, so `isSwipeable` will be undefined on the server and then set on the client - so as is in the example @OlgaLyubin mentioned, aria-hidden will not be set. This will also be the case for AMP - so it's something to think about for that too

DenisHdz replied:
As discussed on Slack, if the `window` is not defined the `useMediaQuery` will return null so `isSwipeable` will be null too, and we won't be applying `aria-hidden: true` and `tabIndex: -1`, therefor it will remain accessible when non-js.

Looking now at AMP.

AlistairGempf replied:
Should this comment be put in the accessibility section of the readme or a comment in the code? It's not an obvious result of the code so might be worth capturing the result.

DenisHdz committed:
Bump top level package.json and update CHANGELOG

DenisHdz committed:
Fix conflicts

DenisHdz committed:
Fix more conflicts

AlistairGempf commented:
The tests look good to me. So other than the questions around amp and disabled js, it all looks good! 👍

DenisHdz committed:
Update Navigation test

DenisHdz committed:
Merge branch 'latest' of github.com:bbc/psammead into swipeable-navigation

DenisHdz committed:
Update snapshots

DenisHdz committed:
Revert latest two snapshots updates

DenisHdz committed:
Revert media placeholder snapshot

DenisHdz committed:
Revert media player snapshot

DenisHdz committed:
Update useMediaQuery hook to check if the window is defined

DenisHdz committed:
Fix conflicts

DenisHdz committed:
Mooooore conflicts

DenisHdz committed:
Pass isSwipeable to Swipeable Navigation as a prop

DenisHdz committed:
Remove WindowWidth and Overflowed hooks

DenisHdz committed:
Split navigation into Canonical and AMP, display always the gradient instead of using JS and add some space to the last list item

DenisHdz committed:
Update tests and README

DenisHdz committed:
Merge branch 'latest' into swipeable-navigation

tochwill commented on packages/components/psammead-navigation/hooks/useMediaQuery.js:
```diff
@@ -0,0 +1,25 @@
+  const [matches, setMatches] = useState(mediaMatches);
+
+  useEffect(() => {
+    if (!isClient) {
```
Does useEffect ever run on the server? I believe it's a no-op

DenisHdz replied:
https://github.com/gatsbyjs/gatsby/issues/13947#issuecomment-490992453
https://github.com/gatsbyjs/gatsby/issues/13947#issuecomment-491214724

tochwill replied:
That second comment is what I mean, `useEffect` doesn't run on the server so we don't need to check if we're on the client?

DenisHdz replied:
Yep, but I have to check it at the top level to set `mediaMatch` as undefined when not on the client? Unless I set it as undefined by default.

tochwill replied:
Yeah sorry - so you need that check, it's just this one inside the useEffect hook that I don't think we need :)

DenisHdz committed:
Update README

DenisHdz committed:
Merge branch 'latest' into swipeable-navigation

DenisHdz committed:
Merge branch 'swipeable-navigation' of github.com:bbc/psammead into swipeable-navigation

DenisHdz committed:
Export swipeable-navigation

DenisHdz committed:
Fix typo

DenisHdz committed:
Fix typo

DenisHdz committed:
Fix conflicts and bump

DenisHdz committed:
Update README

DenisHdz committed:
Fix more conflicts

DenisHdz committed:
Bump version

DenisHdz committed:
Another conflict

tochwill commented on packages/components/psammead-navigation/src/index.jsx:
```diff
@@ -90,6 +80,44 @@ const StyledLink = styled.a`
+  position: relative;
+  z-index: 2;
+
+  @media (max-width: ${GEL_GROUP_2_SCREEN_WIDTH_MAX}) {
```
![Screenshot 2019-11-21 at 14 11 09](https://user-images.githubusercontent.com/6690955/69346050-0d741700-0c6a-11ea-9191-847048644d6d.png)

This doesn't quite look right - would it be easier to keep the hover/focus states as they are, and just add the extra padding to the the list element instead of the link? Or the whole `<ul>` if that works?

DenisHdz replied:
Using `margin` instead of `padding` solves the issue :)

tochwill replied:
Ok cool! :) Could you just do
```
&:last-child {
      ${({ dir }) => css`
        margin-${dir === 'ltr' ? 'right' : 'left'}: ${GRADIENT_WIDTH}; 
  `}
}
```
and not have to override StyledLink styles at all?

DenisHdz committed:
Replace last list item padding with margin

DenisHdz committed:
Remove condition from inside the useEffect

DenisHdz committed:
Apply marign-right/left to last list item not link

AlistairGempf commented on CHANGELOG.md:
```diff
@@ -3,9 +3,10 @@
 | 2.0.10 | [PR#2666](https://github.com/bbc/psammead/pull/2666) Talos - Bump Dependencies - @bbc/psammead-image-placeholder, @bbc/psammead-media-indicator |
 | 2.0.9 | [PR#2657](https://github.com/bbc/psammead/pull/2657) Update codeowners |
-| 2.0.8 | [PR#2663](https://github.com/bbc/psammead/pull/2663) Talos - Bump Dependencies - @bbc/psammead-assets |
+| 2.0.8 | [PR#2663(https://github.com/bbc/psammead/pull/2663) Talos - Bump Dependencies - @bbc/psammead-assets |
```
```suggestion
| 2.0.8 | [PR#2663](https://github.com/bbc/psammead/pull/2663) Talos - Bump Dependencies - @bbc/psammead-assets |
```

DenisHdz committed:
Update CHANGELOG.md

Co-Authored-By: AlistairGempf <43134742+AlistairGempf@users.noreply.github.com>

AlistairGempf reviewed (approved):
I think this all looks good. Good job with this!

sadickisaac commented on packages/components/psammead-navigation/README.md:
```diff
@@ -39,52 +44,78 @@ The `@bbc/psammead-navigation` package is a set of two components, `NavigationUl
+      <NavigationLi
+        url="/"
+        script={latin}
+        active="true"
```
```suggestion
        active
```

sadickisaac commented on packages/components/psammead-navigation/README.md:
```diff
@@ -39,52 +44,78 @@ The `@bbc/psammead-navigation` package is a set of two components, `NavigationUl
 
 We have also added visually hidden text to let the user know which item in the Navigation is the current page. Note the use of visually hidden text here is due to lack of support at this time for the aria-current page attribute. Also note the use of `role="text"` to stop splitting in VoiceOver.
 
+In the screen reader UX only the menu button and its content should be available to assistive technology. The exposed swipeable navigation should be hidden. To achieve this we add `aria-hidden:true` to the exposed swipeable navigation so that this is not expose to these users and also add `tabindex=-1` to the links contained within this to remove them from the tab order.
```
Can we replace the word `exposed` with `visible`?

DenisHdz replied:
I think `exposed` is more appropriate in this context because the navigation will be still visible, but not exposed to screen readers.

sadickisaac replied:
Aaah ok. I was misinterpreting what was meant by the statement because of the use of `exposed` and `hidden` in the same sentence. Would it be better to restructure the sentence a little.

```suggestion
In the screen reader UX only the menu button and its content should be available to assistive technology. The swipeable navigation is exposed to assistive technologies and  should be hidden. To achieve this we add `aria-hidden=true` to the swipeable navigation. We also add `tabindex=-1` to the links contained within this to remove them from the tab order.
```

DenisHdz replied:
I renamed it in a different way, hope this is more clear.

sadickisaac commented on packages/components/psammead-navigation/hooks/useMediaQuery.js:
```diff
@@ -0,0 +1,21 @@
+  const [matches, setMatches] = useState(mediaMatches);
+
+  useEffect(() => {
+    const handler = e => setMatches(e.matches);
```
`e` is not very informative. Is it `event` or `mediaMatch`.

sadickisaac commented on packages/components/psammead-navigation/src/index.jsx:
```diff
@@ -109,43 +125,94 @@ const CurrentLink = ({ children: link, script, currentPageText }) => (
-      >
-        <CurrentLink script={script} currentPageText={currentPageText}>
+}) => {
+  const tabIndex = isSwipeable && { tabIndex: -1 };
```
```suggestion
  const tabIndex = isSwipeable &&  -1;
```
`tabIndex` is already defined as a variable. We don't need to create an object with `tabIndex` as a key again IMO.

sadickisaac replied:
@DenisHdz can we revert this change.  Applying my suggestion leads to console warnings which we have been trying to get rid off in both `psammead` and `simorgh`

<img width="553" alt="Screenshot 2019-11-24 at 16 52 27" src="https://user-images.githubusercontent.com/48688870/69495754-74e4cd80-0edb-11ea-84de-eb3bf5301cff.png">

DenisHdz replied:
Sure. I initially did it in that way to just add `tabIndex` to the DOM when `isSwipeable` is true.

sadickisaac commented on packages/components/psammead-navigation/src/index.jsx:
```diff
@@ -109,43 +125,94 @@ const CurrentLink = ({ children: link, script, currentPageText }) => (
+          service={service}
+          currentLink
+          {...props}
+          {...tabIndex}
```
```suggestion
          tabIndex={tabIndex}
```

sadickisaac commented on packages/components/psammead-navigation/src/index.jsx:
```diff
@@ -109,43 +125,94 @@ const CurrentLink = ({ children: link, script, currentPageText }) => (
+          href={url}
+          script={script}
+          service={service}
+          {...tabIndex}
```
```suggestion
          tabIndex={tabIndex}
```

DenisHdz committed:
Apply README suggestions

DenisHdz committed:
Merge branch 'swipeable-navigation' of github.com:bbc/psammead into swipeable-navigation

DenisHdz committed:
Apply CR feedback

DenisHdz committed:
Merge branch 'latest' of github.com:bbc/psammead into swipeable-navigation

Bopchy reviewed (approved):
Looks really good

sadickisaac commented on packages/components/psammead-navigation/src/index.test.jsx:
```diff
@@ -2,34 +2,62 @@ import React from 'react';
+describe('Swipeable Navigation', () => {
+  window.matchMedia = jest.fn().mockImplementation(query => {
+    return {
+      matches: false,
```
```suggestion
      matches: true,
```
For the swipeable  navigation should the `matches` evaluate to true?

sadickisaac commented on packages/components/psammead-navigation/hooks/useMediaQuery.test.js:
```diff
@@ -0,0 +1,17 @@
+    const { result } = renderHook(() => useMediaQuery('(max-width: 600px)'));
+    expect(result.current).toBe(true);
+  });
+});
```
This seems to not test the `useMediaQuery` hook. If I delete the `useEffect` block in `useMediaQuery` hook the test still pass.  This test seem to test `useState given an initial state it will return the set state`. For this test to be complete we would need to find a way to trigger a window resize to make use of the `useEffect` block.

DenisHdz replied:
You are right. Since I didn't find a way to trigger the window resize from the test, what I have done is to put 600px in the mock, so when I call the `useMediaQuery` hook with `(max-width: 600px)` I expect it to return true.

AlistairGempf replied:
We can check that the `addListener` function has been called. We may also be able to check the effects of the function are what we expect. Looking into that as well.

```jsx
    const mockAddListener = jest.fn();
    const mockRemoveListener = jest.fn();

    window.matchMedia = jest.fn().mockImplementation(query => {
      return {
        matches: query === '(max-width: 600px)',
        addListener: mockAddListener,
        removeListener: mockRemoveListener,
      };
    });

    const { result } = renderHook(() => useMediaQuery('(max-width: 600px)'));
    expect(result.current).toBe(true);
    expect(window.matchMedia.addListener).toBeCalled();
```

AlistairGempf replied:
Here's a version of test that also checks the `addListener` function updates the `result.current`. Note that this should probably be at least two separate tests.

```jsx
  it('it should test that the media query matches', async () => {
    const mockAddListener = jest.fn();
    const mockRemoveListener = jest.fn();

    window.matchMedia = jest.fn().mockImplementation(query => {
      return {
        matches: query === '(max-width: 600px)',
        addListener: mockAddListener,
        removeListener: mockRemoveListener,
      };
    });

    const { result } = renderHook(() => useMediaQuery('(max-width: 600px)'));
    expect(result.current).toBe(true);
    expect(mockAddListener).toBeCalled();

    act(() => mockAddListener.mock.calls[0][0]({ matches: false }));
    expect(result.current).toBe(false);
    act(() => mockAddListener.mock.calls[0][0]({ matches: true }));
    expect(result.current).toBe(true);
  }
```

DenisHdz replied:
Done. Thanks for the suggestions :)

DenisHdz committed:
Revert tabIndex change

DenisHdz committed:
Change matches to true

greenc05 commented on packages/components/psammead-navigation/CHANGELOG.md:
```diff
@@ -3,6 +3,7 @@
 <!-- prettier-ignore -->
 | Version | Description |
 |---------|-------------|
+| 5.0.0-alpha.1 | [PR#2577](https://github.com/bbc/psammead/pull/2577) Add swipeable navigation |
```
Can we call this scrollable rather than swipeable please?

greenc05 commented on packages/components/psammead-navigation/README.md:
```diff
@@ -8,6 +8,10 @@ This component is currently tagged as alpha and is not suitable for production u
 
+## Exports
+
+`/swipeable` - Makes the navigation to be swipeable under 600px.
```
Re name to 'scrollable' please - we aren't adding any swiping gestures?

greenc05 replied:
Same goes for anywhere else we've used the word swipeable...

greenc05 commented on packages/components/psammead-navigation/README.md:
```diff
@@ -39,52 +44,78 @@ The `@bbc/psammead-navigation` package is a set of two components, `NavigationUl
 
 The `Navigation` is designed to show a navigation bar on `index` pages, which will show all sections on a site. If there are too many items to fit on one line, the items will wrap to the next lines.
 
+On the other hand, with `CanonicalSwipeableNavigation` or `AmpSwipeableNavigation` we can make the list to remain on one line and to be horizontally swipeable to allow access to further links, under 600px.
```
'horizontally swipeable to' should be 'horizontally scrollable to'

tochwill commented on packages/components/psammead-navigation/hooks/useMediaQuery.js:
```diff
@@ -0,0 +1,21 @@
+
+    mediaMatch.addListener(handler);
+    return () => mediaMatch.removeListener(handler);
+  });
```
Is this going to re-run useEffect after every render? As you're just adding a listener when the component mounts, could you add `[]` as the second argument to useEffect?

greenc05 commented on packages/components/psammead-navigation/README.md:
```diff
@@ -39,52 +44,78 @@ The `@bbc/psammead-navigation` package is a set of two components, `NavigationUl
 
 We have also added visually hidden text to let the user know which item in the Navigation is the current page. Note the use of visually hidden text here is due to lack of support at this time for the aria-current page attribute. Also note the use of `role="text"` to stop splitting in VoiceOver.
 
+In the screen reader UX only the menu button and its content should be available to assistive technology, for which reason the swipeable navigation should be hidden. To achieve this we add `aria-hidden:true` to the exposed swipeable navigation so that this is not visible to these users and also add `tabindex=-1` to the links contained within this to remove them from the tab order.
```
In the screen reader UX only the menu button and its content should be available to assistive technology, **meaning the scrollable navigation will be hidden**. To achieve this we add `aria-hidden:true` to the exposed **scrollable** navigation so that this is not visible to these users and also add `tabindex=-1` to the links contained within this to remove them from the tab order.

DenisHdz committed:
Replace Swipeable with Scrollable

DenisHdz committed:
Apply Alistair feedback

DenisHdz committed:
Merge branch 'latest' of github.com:bbc/psammead into swipeable-navigation

OlgaLyubin reviewed (commented):
There's still a folder named swipeable
<img width="669" alt="Screenshot 2019-11-25 at 15 13 11" src="https://user-images.githubusercontent.com/17802955/69552302-234d4900-0f96-11ea-8f1a-b0693f12f75a.png">

DenisHdz committed:
Rename folder

tochwill commented on packages/components/psammead-navigation/hooks/useMediaQuery.test.js:
```diff
@@ -0,0 +1,35 @@
+    expect(mockAddListener).toBeCalled();
+  });
+
+  it('should add a listener that updates the state', async () => {
```
Just wondering why these functions are async?

DenisHdz committed:
Remove async

sadickisaac commented on packages/components/psammead-navigation/src/ScrollableNavigation/index.jsx:
```diff
@@ -0,0 +1,75 @@
+CanonicalScrollableNavigation.defaultProps = { dir: 'ltr' };
+
+export const AmpScrollableNavigation = ({ children, dir }) => (
+  <StyledScrollableNav dir={dir}>{children}</StyledScrollableNav>
```
Should we have a `arria-hidden` here?

DenisHdz replied:
No, since we can not use JS in AMP to determine if it should use `aria-hidden` or not, the scrollable nav will be fully available to keyboard users via the tab key and to screen reader users.

DenisHdz committed:
Fix conflicts

DenisHdz committed:
Merge branch 'latest' into swipeable-navigation

DenisHdz committed:
Fix conflicts

DenisHdz committed:
Merge branch 'swipeable-navigation' of github.com:bbc/psammead into swipeable-navigation

DenisHdz committed:
Merge branch 'latest' into swipeable-navigation

DenisHdz committed:
Fix conflicts

DenisHdz committed:
Merge branch 'swipeable-navigation' of github.com:bbc/psammead into swipeable-navigation

DenisHdz committed:
Merge branch 'latest' into swipeable-navigation

jamesbrumpton commented:
lgtm

DenisHdz committed:
Fix conflicts

DenisHdz committed:
Merge branch 'swipeable-navigation' of github.com:bbc/psammead into swipeable-navigation

DenisHdz committed:
Revert brand changes

DenisHdz committed:
Conflicts again
